### ISSUE TITLE
not_null's dereference operator does not work for smart pointers with c++11

### ISSUE THREAD

**Issue by petamas** [martinmoene/gsl-lite#184]
The second and third test case does not compile when gcc is set to C++11 mode:

```
using point = std::pair<int, int>;

TEST(GslNotNull, NNRawDereference)
{
	not_null<point*> nnp(new point(1, 3));
	EXPECT_EQ(*nnp, std::make_pair(1,3));
}

TEST(GslNotNull, NNUniqueDereference)
{
	not_null<unique_ptr<point>> nnp(new point(1, 3));
    EXPECT_EQ(*nnp, std::make_pair(1,3));
}

TEST(GslNotNull, NNSharedDereference)
{
	not_null<shared_ptr<point>> nnp(new point(1, 3));
    EXPECT_EQ(*nnp, std::make_pair(1,3));
}
```

See: https://gcc.godbolt.org/z/TtEth-

The problem is that operator* is defined using decltype(auto), which is available from C++14. It should be replaced with something like `decltype(*declval<T>())` or with something along the lines of Dropbox nn:
```
namespace nn_detail {
template <typename T> struct element_type { using type = typename T::element_type; };
template <typename Pointee> struct element_type<Pointee *> { using type = Pointee; };
}
// ...
using element_type = typename nn_detail::element_type<PtrType>::type;
// ...
element_type & operator*()  const { return *ptr; }
```

The declval approach is problematic, because it doesn't work with C++03; in contrast, the nn-style approach compiles with C++03, but requires smart pointers to have an embedded `element_type` typedef.

**Comment by mbeutel** [comment]
Thank you for the analysis. Yes, I think we can do better than defining `operator *()` for C++14 only.

Perhaps we can use `PtrT::element_type` if it exists, and `decltype(*declval<T>())` otherwise (if C++11).

**Comment by petamas** [comment]
I was thinking the reverse: `if(C++11) { use decltype } else { use PtrT::element_type }` - this way we only add restrictions to PtrT in pre-modern C++, where smart pointers are used much less. Also, checking whether PtrT::element_type exists can be quite tricky, so GSL's code would be easier to understand this way.

**Comment by mbeutel** [comment]
We wouldn't be adding restrictions to `PtrT` in C++11 because we'd check if `PtrT::element_type` exists, and only use it if it does. By trying `PtrT::element_type` first, we keep the return type selection stable with respect to the C++ language version (though admittedly `PtrT::element_type` is unlikely to differ from `decltype(*declval<T>())`).

**Comment by petamas** [comment]
I think I understand your rationale; I'll try to whip up something next week.

**Comment by petamas** [comment]
@mbeutel: Did play around a little bit on CE, made this: https://godbolt.org/z/qJ6A7B

What do you think? Also, should we add a static_assert to check if element_type and the decltype-based approach give the same result, or leave it unchecked? (My "WeirdPtr" example would fail the check.)

**Comment by mbeutel** [comment]
Looks good. Note that we already define [`void_t`](https://github.com/martinmoene/gsl-lite/blob/master/include/gsl/gsl-lite.hpp#L567) in *gsl-lite*, so can use that instead of `conditional<>`.

Do you want to prepare a PR? If not I'll try to get it done later this week.

**Comment by petamas** [comment]
I'll prepare a PR, I need the practice. :) I haven't seen void_t yet, nice to know, I'll use it. :)

**Comment by petamas** [comment]
Note: gsl::std17::void_t does not work with gcc 4: https://godbolt.org/z/akaSaH

I think the reason is this:
> Until [CWG 1558](http://open-std.org/JTC1/SC22/WG21/docs/cwg_defects.html#1558) (a C++14 defect), unused parameters in [alias templates](https://en.cppreference.com/w/cpp/language/type_alias) were not guaranteed to ensure SFINAE and could be ignored, so earlier compilers require a more complex definition of void_t, such as
> `template<typename... Ts> struct make_void { typedef void type;};`
> `template<typename... Ts> using void_t = typename make_void<Ts...>::type;`
Source: https://en.cppreference.com/w/cpp/types/void_t

If I use my_void_t (defined as cppreference.com says), it works. I'll update GSL-Lite's definition of void_t to match this.

**Comment by mbeutel** [comment]
Fixed by #192, thank you.
